[Abandoned] Port of Vim's `flatten`

cbarrete opened the pull request:
This ports patches 8.2.0935 and 8.2.0937.

Fixes https://github.com/neovim/neovim/issues/12541.

Commit:
Port Vim's `flatten` function

This ports patches 8.2.0935 and 8.2.0937.

Commit:
Fix lint warnings

Commit:
Fix another lint warning

janlazo commented:
https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#quickstart
None of the commit descriptions in this PR follow the format used by https://github.com/neovim/neovim/blob/master/scripts/vim-patch.sh so that ported patches can be tracked by Neovim tooling. State your intentions if your patches are full ports of Vim's patches `flatten` or not. It's ok if they're not but always include the Vim commit hash so another developer can finish porting the patches. Update the PR title so that the bot can tag your PR for convenience.

Commit:
patch 8.2.0935: flattening a list with existing code is slow

Problem:    Flattening a list with existing code is slow.
Solution:   Add flatten(). (Mopp, closes #3676)

Commit:
patch 8.2.0937: asan failure in the flatten() test

Problem:    Asan failure in the flatten() test.
Solution:   Free the flattened list.

cbarrete commented:
This is a full port indeed. Is there something else that I should do for this PR to be ready?

teto committed:
doc: fix wordcount description

was fixed in vim patch c572da5f67aa5cdbbc127fc6f1d0a42e38468325
and forgot in corresponding vim patch 8c3377ee76e5cc4f5a4228f216bde1d03b496d69.

janlazo commented:
Is it difficult to rebase this PR so that the `vim-patch.sh` generates the commit description? I prefer to not add these patches manually in a version.c PR like https://github.com/neovim/neovim/pull/12524 and hunt down related vim patches like `Termdebug` in https://github.com/neovim/neovim/pull/8364.

Tests should pass before this PR can be merged. Sourcehut failures are expected but Travis build should not fail, at least for the first stage. Appveyor builds take a long to finish so but most tests should pass, aside from the flaky async tests like jobs and terminal.

erw7 commented on src/nvim/eval/funcs.c:
```diff
@@ -2157,6 +2157,43 @@ static void f_expandcmd(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
+/*
+ * "flatten(list[, {maxdepth}])" function
+ */
```
There is an error in `clint` as shown in the log below. Please amend this to a comment in `//` format.

https://travis-ci.org/github/neovim/neovim/jobs/703247240#L2373

cbarrete replied:
Sure, I just followed the local convention. I'll change that.

cbarrete commented:
@janlazo I'll just close this PR and start a new one using the script then, it'll likely be easier and cleaner.

cbarrete commented:
Changing the name of the PR to not interfere with the bot.
